File: core/src/kr/edoli/imview/ui/histogram/HistogramPanel.java
package kr.edoli.imview.ui.histogram;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import kr.edoli.imview.ui.UIFactory;
import kr.edoli.imview.ui.drawable.ColorDrawable;
import kr.edoli.imview.ui.res.Colors;

/**
 * Created by daniel on 16. 2. 27.
 */
public class HistogramPanel extends Table {
    public HistogramPanel() {
        final HistogramViewer histogramViewer = new ColorHistogramViewer();

        add(histogramViewer).expand().fill().row();

        Table row = new Table();

        String[] texts = new String[] {
                "red", "green", "blue"
        };

        for (int i = 0; i < histogramViewer.histogramNumber(); i++) {
            final Button checkBox = UIFactory.checkBox(texts[i]);
            final int index = i;

            checkBox.addListener(new ChangeListener() {
                @Override
                public void changed(ChangeEvent event, Actor actor) {
                    histogramViewer.setShow(index, checkBox.isChecked());
                }
            });
            checkBox.setChecked(histogramViewer.isShow(i));
            row.add(checkBox).expandX().fillX().height(20).pad(4);
        }

        add(row).expandX().fillX();

        background(new ColorDrawable(Colors.background));
    }
}

File: desktop/src/main/kotlin/kr/edoli/imview/ui/panel/histogram/HistogramPanel.kt
package kr.edoli.imview.ui.panel.histogram

import com.badlogic.gdx.scenes.scene2d.ui.Table
import kr.edoli.imview.ImContext
import kr.edoli.imview.image.split
import kr.edoli.imview.ui.Panel
import kr.edoli.imview.ui.UIFactory
import kr.edoli.imview.util.Histogram
import org.opencv.core.CvType

class HistogramPanel : Panel(false) {

    val histogramActor = HistogramWidget()
    val buttons = Table()

    init {
        add(histogramActor).expandX().fillX().height(128f)
        row()
        add(buttons).expandX().fillX()
    }

    fun updateHistogram() {
        val spec = ImContext.mainImageSpec.get() ?: return
        val mat = spec.mat

        val minMax = ImContext.imageMinMax.get()
        val minValue = minMax.first
        val maxValue = minMax.second

        val numBin = 256

        val histograms = mat.split().map { singleMat ->
            Histogram(numBin).apply {
                computeHistMat(singleMat, minValue, maxValue)
            }
        }

        histogramActor.histograms.clear()
        histogramActor.histograms.addAll(histograms)

        buttons.clearChildren()
        val names = when (mat.channels()) {
            1 -> arrayOf("Gray")
            3 -> arrayOf("Red", "Green", "Blue")
            4 -> arrayOf("Red", "Green", "Blue", "Alpha")
            else -> arrayOf()
        }

        (names.indices).forEach { index ->
            buttons.add(UIFactory.createTextButton(names[index]) {
                histogramActor.isShow[index] = it.isChecked
            }.apply {
                isChecked = true
                style = UIFactory.textToggleButtonStyle
            })
        }
    }
}